### PR TITLE
Refactor evaluator to cache getters by reference, not hash code

### LINKED ISSUE #140: Change observers not called for some changes
I've hit a very weird bug in my app. Certain orderings of changes to my store result in change observers not being called.

I'm concerned that maybe I'm running into #86 

I've managed to create a minimal reproduction.

```js
var Nuclear = require("nuclear-js"); // v1.1.1

var SET_YEAR_GROUP = "SET_YEAR_GROUP";
var LOADED = "LOADED";

function runTest(value) {

  var store = Nuclear.Store({
    getInitialState: function() {
      return Nuclear.toImmutable({
        yearGroup: 0,
        shouldLoaded: true
      });
    },

    initialize: function() {
      this.on(SET_YEAR_GROUP, setYearGroup);
      this.on(LOADED, loaded);
    }
  });

  function setYearGroup(store, payload) {
    return store
      .set("yearGroup", payload.yearGroup)
      .set("shouldLoad", true);
  }

  function loaded(store) {
    return store.set("shouldLoad", false);
  }

  var reactor = Nuclear.Reactor();

  reactor.registerStores({ uiStore: store });

  var output = [];
  // Record changes to yearGroup
  reactor.observe(["uiStore", "yearGroup"], function(y) { output.push(y) });

  reactor.dispatch(SET_YEAR_GROUP, {yearGroup: 6});
  reactor.dispatch(LOADED);
  reactor.dispatch(SET_YEAR_GROUP, {yearGroup: value});

  return output;
}

var output1 = runTest(1);
console.log(output1); // OK - logs [ 6, 1 ]

var output5 = runTest(5);
console.log(output5); // WRONG! Logs [ 6 ] instead of [ 6, 5 ] !

// Removing the reactor.dispatch("LOADED"); call seems to fix things, but why?!
```

Is this a bug in Nuclear?

### PR DESCRIPTION
**Pull request by jordangarcia**
This fixes #140.

My concern here is about increased memory usage, will run a few other benchmarks to make sure this is okay.

@bhamodi @andrewdavey

### CODE REVIEW

File: test.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
+var Nuclear = require("./dist/nuclear"); // v1.1.1
```
Comment (bhamodi) [inline test.js]: Did you mean to commit this file? If so, it should be moved to the `tests` directory.

File: tests/reactor-tests.js [hunk 2]
```diff
@@ -1144,4 +1144,49 @@ describe('Reactor', () => {
       }).not.toThrow()
     })
   })
+
+  describe('issue #140 - change observer error case dealing with hashCode collisions', () => {
```
Comment (bhamodi) [inline tests/reactor-tests.js]: I don't think we need the issue number in the test description here.

**Comment by bhamodi** [comment]
2 minor comments above to address, but other than that, it LGTM. I'd also be interested in seeing the change in performance. Could you post some stats?

**Comment by andrewdavey** [comment]
Can this be merged and a new build released? Thanks.

**Comment by jordangarcia** [comment]
Unfortunately still needs a bit of work. The more aggressive caching mechanism lead to too many things being held on to by reference which caused memeory leaks.

I am working on refactoring this in a different branch now.  Sorry for the inconvenience.

**Comment by bhamodi** [comment]
Any updates on this @jordangarcia?

**Comment by jakecraige** [comment]
We also ran into this today and it took a good hour or so trying to debug. Would love to see this change come through/

**Comment by bhamodi** [comment]
This has been done in v1.2.0.

**Pull request closed without merging**